### PR TITLE
feat(pipeline): prune receipts based on log emitters

### PR DESCRIPTION
**Pull request by joshieDo**
This allows us to only include the deposit contract receipts, alongside whatever the user chooses to keep (ref https://github.com/paradigmxyz/reth/issues/2753) . (for pipeline only)

### CODE REVIEW

File: crates/primitives/src/prune/mod.rs [hunk 1]
```diff
@@ -7,3 +7,38 @@ pub use checkpoint::PruneCheckpoint;
...
+    ///
+    /// The [`BlockNumber`] key of the map should be viewed as `PruneMode::Before(block)`.
+    pub fn flatten(
+        &self,
+        tip: BlockNumber,
+    ) -> Result<BTreeMap<BlockNumber, Vec<&Address>>, PrunePartError> {
+        let mut map = BTreeMap::new();
+        for (mode, address) in self.0.iter() {
+            let block = mode
+                .prune_target_block(tip, 0, PrunePart::ContractLogs)?
```
Comment (shekhirin) [inline crates/primitives/src/prune/mod.rs:36]: I think we need receipts to unwind in blockchain tree
```suggestion
                .prune_target_block(tip, 128, PrunePart::ContractLogs)?
```

File: crates/primitives/src/prune/mod.rs [hunk 2]
```diff
@@ -7,3 +7,38 @@ pub use checkpoint::PruneCheckpoint;
...
+                .map(|(block, _)| block)
+                .unwrap_or_default();
```
[the remark this answers is not in the document]
Comment (joshieDo) [reply crates/primitives/src/prune/mod.rs:42]: added some docs, does it make sense ?

File: crates/primitives/src/prune/mod.rs [hunk 3]
```diff
@@ -6,4 +6,43 @@ mod target;
 pub use checkpoint::PruneCheckpoint;
 pub use mode::PruneMode;
 pub use part::{PrunePart, PrunePartError};
-pub use target::PruneModes;
+pub use target::{PruneModes, MINIMUM_PRUNING_DISTANCE};
+
```
Comment (mattsse) [inline crates/primitives/src/prune/mod.rs:10]:
```suggestion
```

File: crates/primitives/src/prune/mod.rs [hunk 4]
```diff
@@ -6,4 +6,43 @@ mod target;
...
+use crate::{Address, BlockNumber};
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeMap;
+
+/// Configuration for pruning receipts not associated with logs emitted by the specified contracts.
+#[derive(Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize)]
+pub struct ContractLogsPruneConfig(pub Vec<(PruneMode, Address)>);
```
Comment (mattsse) [inline crates/primitives/src/prune/mod.rs:17]: should this be a btreemap<address, PruneMode> instead?

File: crates/primitives/src/prune/mode.rs [hunk 5]
```diff
@@ -14,6 +14,42 @@ pub enum PruneMode {
...
 
+impl PruneMode {
+    /// Returns block up to which variant pruning needs to be done, inclusive, according to the
+    /// provided tip.
+    pub fn prune_target_block(
+        &self,
+        tip: BlockNumber,
+        min_blocks: u64,
+        prune_part: PrunePart,
+    ) -> Result<Option<(BlockNumber, PruneMode)>, PrunePartError> {
```
Comment (mattsse) [inline crates/primitives/src/prune/mode.rs:25]: can we add a sanity test for this, code looks correct afaict, just to be sure

File: crates/primitives/src/prune/mode.rs [hunk 6]
```diff
@@ -14,6 +14,42 @@ pub enum PruneMode {
...
+        Ok(match self {
```
Comment (mattsse) [inline crates/primitives/src/prune/mode.rs:26]: style nit: return result in two steps to get rid of the large Ok( wrapper

**Review by mattsse — CHANGES_REQUESTED** [review]
nit and question re format

File: crates/primitives/src/prune/mod.rs (hunk 4 above)
Comment (joshieDo) [reply crates/primitives/src/prune/mod.rs:16]: took a stylistic choice on how its represented on the `toml` file, but I'm open to change:

btreemap:
```
[parts.contract_logs_filter.0x00000000219ab540356cbb839cbe05303d7705fa]
before = 11052984

[parts.contract_logs_filter.0xf08a50178dfcde18524640ea6618a1f965821715]
before = 3802888
```

vec:
```
contract_logs_filter = [
    [
        { before = 11052984 },
        "0x00000000219ab540356cbb839cbe05303d7705fa",
    ],
    [
        { before = 3802888 },
        "0xf08a50178dfcde18524640ea6618a1f965821715",
    ],
]
```

Comment (joshieDo) [reply crates/primitives/src/prune/mod.rs:16]: @shekhirin

Comment (shekhirin) [reply crates/primitives/src/prune/mod.rs:16]: I'd rather go with btreemap, vec looks like a json inside toml for me

File: crates/primitives/src/prune/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (shekhirin) [reply crates/primitives/src/prune/mod.rs:41]: ah, yes, thanks!

**Review by shekhirin — APPROVED** [review]
LGTM modulo Matt's comments

**Comment by joshieDo** [comment]
testing on sepolia by including only receipts from the deposit contract and another random contract

**Comment by joshieDo** [comment]
worked fine. receipt table ended the pipeline with ~15M keeping all receipts from the last 128 blocks, and further only the ones (2 contracts) that were present in the configuration

**Review by mattsse — APPROVED** [review]
tests look good,

lgtm

**Pull request merged**